tensorflow/c/eager/dlpack.cc

```diff
@@ -0,0 +1,327 @@
+                                                    TF_Status* status) {
+  switch (ctx.device_type) {
+    case DLDeviceType::kDLCPU:
+      return "CPU:0";
+    case DLDeviceType::kDLGPU:
+      return absl::StrCat("GPU:", ctx.device_id);
+    default:
+      return absl::nullopt;
+  };
+}
+TF_DataType TfDataTypeFormDlDataType(const DLDataType& dtype,
+                                     TF_Status* status) {
+  TF_DataType tf_dtype;
+  switch (dtype.code) {
+    case DLDataTypeCode::kDLUInt:
+      switch (dtype.bits) {
+        case 8:
+          tf_dtype = TF_DataType::TF_UINT8;
+          break;
+        case 16:
+          tf_dtype = TF_DataType::TF_UINT16;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_UINT32;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_UINT64;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported UInt bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLInt:
+      switch (dtype.bits) {
+        case 8:
+          tf_dtype = TF_DataType::TF_INT8;
+          break;
+        case 16:
+          tf_dtype = TF_DataType::TF_INT16;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_INT32;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_INT64;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported Int bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLFloat:
+      switch (dtype.bits) {
+        case 16:
+          tf_dtype = TF_DataType::TF_HALF;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_FLOAT;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_DOUBLE;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported Float bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLBfloat:
+      switch (dtype.bits) {
+        case 16:
+          tf_dtype = TF_DataType::TF_BFLOAT16;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported BFloat bits: ", dtype.bits);
+      }
+      break;
+    default:
+      status->status = tensorflow::errors::InvalidArgument(
+          "Unsupported Type Codes: ", dtype.code);
+  }
+
+  return tf_dtype;
+}
+
+void DeallocatorWrapperFunc(void* data, size_t len, void* dlmt_vptr) {
+  DLManagedTensor* dlmt = static_cast<DLManagedTensor*>(dlmt_vptr);
+  dlmt->deleter(const_cast<DLManagedTensor*>(dlmt));
+}
+
+bool IsValidStrideCompactRowMajorData(int64_t* shape_arr, int64_t* stride_arr,
+                                      int ndim) {
+  for (int i = ndim - 2; i >= 0; --i) {
+    if (stride_arr[i] != shape_arr[i + 1] * stride_arr[i + 1]) {
+      return false;
+    };
+  }
+  return true;
```

Comment:
Don't you also need to check that `stride_arr[ndim-1]` is `1`?

Reply:
Added